ledger/src/blockstore.rs

```diff
@@ -5846,31 +5825,23 @@ pub mod tests {
             .unwrap();
 
         let slot_meta = blockstore.meta(0).unwrap().unwrap();
-        // Corrupt shred by making it too large
-        let mut shred5 = shreds[5].clone();
-        shred5.payload.push(10);
-        shred5.data_header.size = shred5.payload.len() as u16;
-        assert!(!blockstore.should_insert_data_shred(
-            &shred5,
-            &slot_meta,
-            &HashMap::new(),
-            &last_root,
-            None,
-            ShredSource::Turbine
-        ));
```

Comment:
moved to `test_sanitize_data_shred`

Reply:
I think these are still useful test cases in case somebody accidentally removes/moves the call to sanitize() in the `should_insert_data_shred` function. 

Same as my comment here: https://github.com/solana-labs/solana/pull/24563#discussion_r856774257, I think if we add test only modifiers to the shred, we could keep these test cases

Reply:
I actually think it is better if we move in the other direction; i.e.
1. Remove all these "test-only" shred methods which can invalidate a shred (`set_slot`, `set_index`, `set_last_in_slot`, `unset_data_complete`, ...):
https://github.com/solana-labs/solana/blob/92ad76773/ledger/src/shred.rs#L577-L597
https://github.com/solana-labs/solana/blob/92ad76773/ledger/src/shred.rs#L639-L663
2. Ensure that all methods which construct or modify a shred do sanitize shred at the end and can only return a sanitized shred (`new_from_serialized_shred`, `new_from_data`, ...).
   * and, add test coverage that each of these methods correctly returns `Err` if invoked with invalid arguments.
   * `new_from_serialized_shred` already does sanitize before return:
   https://github.com/solana-labs/solana/blob/92ad76773/ledger/src/shred.rs#L357-L360
3. _Remove_ the calls to `shred.sanitize()` from anywhere else outside of `ledger::shred` module.
   * `Shred::sanitize` should in fact be a _private_ function to only `Shred` itself.

(1) and (2) guarantee that if you have a shred it is already sanitized and no need to worry about that anymore. So any call to `Shred::sanitize` outside of `shred.rs` is superfluous.
 
Blockstore, window-service, etc then can only worry about checking shreds properties which do not invalidate shreds by themselves; e.g. if the shred is after the latest root or not, or if it has the right shred-version, etc.

The downside of allowing invalid shreds to fly around but then scattering `shred.sanitize()` all over the code-base is that:
1. It is very easy to forget `shred.sanitize()` on some necessary call-path if invalid shreds are a possibility everywhere. 
2. We will end up calling `shred.sanitize()` (or repeating the same checks as in `shred.sanitize()`) on the same shred several times during its processing lifetime, which is not free; e.g.
   * once in the window service when shreds are de-serialized from the packets payload.
   * then again in `should_retransmit_and_persist`:
   https://github.com/solana-labs/solana/blob/38bdb401a/core/src/window_service.rs#L193-L198
   * then again in `should_insert_data_shred` in blockstore:
   https://github.com/solana-labs/solana/blob/38bdb401a/ledger/src/blockstore.rs#L1446-L1479
   * ...

Reply:
cool, that makes sense